Sources/FluentKit/Enum/OptionalEnumProperty.swift

```diff
@@ -79,6 +79,8 @@ extension OptionalEnumProperty: AnyDatabaseProperty {
     public func input(to input: DatabaseInput) {
         if let value = self.value {
             input.set(value.map { .enumCase($0.rawValue) } ?? .null, at: self.field.key)
+        } else {
+            input.set(.null, at: self.field.key)
```

Comment:
IIRC, this is wrong - the result will be that the enum property's value is _always_ included in `UPDATE` queries, even if it hasn't changed, which breaks various things. I feel like the actual issue is in the setter for `value`, but I can't prove it until I tinker with the code a bit (and I may be wrong). Can you add a unit test that checks that the enum's value is _not_ unconditionally set to `nil` even if there's no corresponding assignment to the property?

Reply:
@gwynee Thanks for taking a look! Re: the unit test, do you mean something like this?

```swift
    func testUpdateDoesntContainNeedlessNullForNilOptionalEnum() throws {
        let db = DummyDatabaseForTestSQLSerializer()
        let workout = Workout(name: "Yoga", distance: nil, effort: nil)
        _ = try workout.create(on: db).wait()
        workout.name = "Goat Yoga"
        _ = try workout.update(on: db).wait()
        XCTAssertEqual(
            db.sqlSerializers.last?.sql.contains(#""effort" = NULL"#),
            false
        )
    }
```

If so, that test fails.  But I guess I don't see the harm. Technically the `SET "effort" = NULL` is redundant in this case (it already _is_ null), but that's no big deal right?  What scenarios "break" due to this?  And, I guess, why are there no tests for those scenarios?

Is there an optimization somewhere where if you call `model.update(on:)` without anything having been changed, that it's a noop?  I could see this "breaking" that case...  But, I guess I would rather have two gnarly bugs fixed and lose that optimization than keep the optimization and bugs.  That said, I _totally_ am in over my head here, and would love it if you tinkered some and found a more correct solution. Thanks again!

Reply:
Having the redundant SET overwrites data. Consider this sequence:
```swift
SomeModel(stringField: "bar", enumField: .foo)
    .save(on: database)
// Some time later, in another route perhaps:
SomeModel.query(on: database).all().flatMap {
    $0.map { $0.stringField = "baz"; $0.update(on: database) }
}.flatten(on: database.eventLoop)
```
This will set the `enumField` value to `NULL` for every single model in the database. This was obviously not the intended effect 😅

Reply:
Hmm... I would have thought in your scenario that `enumField` would only be set to `NULL` for any models where it was already `nil`.

Reply:
Unfortunately, no - `AnyDatabaseProperty.input(to:)` is invoked for every property on the model during query generation. For optional properties (and this is one of the reasons the `OptionalField/Parent/Child/EnumProperty` variants exist), the value is a _double_-optional; `.none` means "no value is set for this property", `.some(.none)` means "this property is explicitly set to `nil`", and `.some(.some(...))` is "this property was explicitly given a value". Additionally, calling `DatabaseInput.set(_:at:)` with any value whatsosever guarantees that the resulting query will have the given field key listed, including setting `NULL` for `nil`. The "no value is set" case is handled by just not calling the method for that property at all - this is how the `Optional*Property` types work. The non-optional properties do the same too; they just never pass `nil` as the value when there is one.

Reply:
(It used to be that `FieldProperty` tried to handle both optional and non-optional fields in a single type, but in the end Tanner found that it caused too much havoc, partially due to Swift's somewhat unintuitive behavior around double-optionals.)

Reply:
@gwynne thanks so much for the in-depth explanation.  I see now that I really truly don't understand the inner workings of this library, and it seems pretty clear that the test suite is not a sufficient safety net for this type of change. 

I don't suppose you'd be willing to take the failing test cases I wrote and work on the correct solution at some point?  I'm very willing to help in any way I can, but after reading your explanation, I don't know if I'm the right person for the job.

The #396 bug is causing me a significant problem in a Vapor app I've just put into production. Because we currently only have a few users (but that number will be growing), as a temporary workaround I'm literally looping over hundreds of records at a time and calling `model.create(on:)` for each one, because I can't do a bulk insert without risking a fatal error. Without this getting fixed, I'm thinking I might need to drop down and write the bulk insert query in raw SQL, which I'm really hoping not to have to do, as I wouldn't feel super confident that I was getting it right for all of the edge cases, and escaping, etc.